### ISSUE TITLE
Support for rendering ANSI-colored text from piped stdin (and use Helix as `$PAGER`)

### ISSUE THREAD

**Issue by NNBnh** [helix-editor/helix#1074]
Although #996 have been merged, support for rendering ANSI-colored text from piped stdin inside fifo buffers is very useful to used as [Kitty terminal](https://sw.kovidgoyal.net/kitty)'s [`scrollback_pager`](https://sw.kovidgoyal.net/kitty/conf/#opt-kitty.scrollback_pager) and will enable Helix to be used as `$PAGER`, `$MANPAGER` and `$GIT_PAGER` without the need of a plugin.

Some Kakoune plugins supporting this feature:
- [Kakpipe](https://github.com/eburghar/kakpipe)
- [kak-ansi](https://github.com/eraserhd/kak-ansi)

Also for using the main editor as `less` or [`bat`](https://github.com/sharkdp/bat#automatic-paging) replacement, Kakoune currently need a dedicated plugin [pager.kak](https://github.com/krornus/pager.kak). Hope that Helix can somehow implement this be defaults.

**Comment by archseer** [comment]
This seems a bit orthogonal to a text editor to me. Why not just use a proper pager?

**Comment by NNBnh** [comment]
> Why not just use a proper pager?

It's the same reason why many seaking for way to use Kakoune and [Vim as pager](https://github.com/rkitover/vimpager):
Pager is just an editor with less feature, way less feature that look and behave ~~badly~~ different, that mean user need to learn and config two different software that do nearly the same jobs: reading document. Having an editor as the pager is way more powerful, consistent and minimal (also, people just love the editor too much and want to use it more).

Btw, Helix has allowed piping from stdin into a buffer, so Helix have *somewhat* already be usable as pager:

```sh
export PAGER="hx"
export MANPAGER="$PAGER"
```

Although Helix currently wasn't able to handle terminal ansi code which lead to open a buffer with a bunch of ansi-characters-escape (this image from Kakoune is for illustration purpose):

![Screenshot_2021-11-11_21-57-34](https://user-images.githubusercontent.com/43980777/141319516-6fb010b6-edb2-4b4f-a280-18187b4cb1f9.png)

We could add a option in the `config.toml` file:

```toml
ansi = true # Default
```

When Helix being piped:
- If `false`, Helix just open a stdin buffer with a bunch of ansi-characters-escape as currently.
- If `true`, Helix remove all ansi-characters-escape and render stdin exactly the same as it's rendered on the shell.

Here are some more demonstrate (achieved using Kakoune with [kak-ansi](https://github.com/eraserhd/kak-ansi)):

[![asciicast](https://asciinema.org/a/43MiUjCNX2SDJy4jyHyijzgP8.svg)](https://asciinema.org/a/43MiUjCNX2SDJy4jyHyijzgP8)

[Kakpipe](https://github.com/eburghar/kakpipe) is also be written in rust, we could took/investegate some of its code.
But still "render stdin exactly the same as it's rendered on the shell" seem difficult, I don't expect this feature to come to Helix even in 2022 base on how long it took to have the same kind of plugin for Kakoune:

> - Kakoune start in 2012
> - **6 year gap...**
> - [kak-ansi start in 2018](https://github.com/eraserhd/kak-ansi/graphs/contributors)

If the dev thinks this feature shouldn't be built in to Helix, please label this issue as plugin idea and the `ansi = true` option will just remove all ansi-characters-escape.

**Comment by kbd** [comment]
I am also a kitty user. It, cleverly, provides scrollback search by piping scrollback to a pager. With its new shell integration, you can also do things like pipe the last command's output to a pager.

The usual `less` isn't great for this, for a bunch of reasons I could enumerate.

I currently use this monstrosity with neovim in my kitty config:

`
scrollback_pager /usr/local/bin/nvim -c "set nonumber nolist showtabline=0 foldcolumn=0 laststatus=0 noshowmode noruler noshowcmd ignorecase smartcase incsearch shortmess+=F" -c "autocmd TermOpen * normal G" -c "map q :qa!<CR>" -c "set clipboard+=unnamedplus" -c "silent write! /tmp/kitty_scrollback_buffer | te echo -n \"$(cat /tmp/kitty_scrollback_buffer)\" && sleep 1000 "
`

There's a [long issues thread](https://github.com/kovidgoyal/kitty/issues/719) for kitty with different people offering up [their recipes](https://github.com/kovidgoyal/kitty/issues/719#issuecomment-952039731) for how to do this.

It is very powerful to pipe your scrollback into your full editor, have all the incremental search/highlight comforts of your normal editor, be able to copy text out of it from the keyboard, and so on. I think it's within the realm of what a terminal-based text editor can/should do. I'd happily replace the above long nvim command with any other editor's "pager mode" that provides the equivalent.

**Comment by ubitux** [comment]
On the topic of pager, neovim proposes this in its documentation: `MANPAGER=nvim +Man!`, which provides color highlights, classic vim bindings, with an extra `q` binding to quit. Pretty handy.

**Comment by Pound-Hash** [comment]
> This seems a bit orthogonal to a text editor to me. Why not just use a proper pager?

I'm not interested in using it as a pager. But I would like to output some content of a man page and then edit it (with helix).

**Comment by bcspragu** [comment]
Another use case for this: code reviews in the terminal. [delta](https://github.com/dandavison/delta) generates syntax highlighted diffs (I believe using the same underlying tooling as [bat](https://github.com/sharkdp/bat)), which look great/substantially like the ones in GitHub, ex:

![2023-10-11_19-03-37](https://github.com/helix-editor/helix/assets/2402010/11768214-a387-474c-880f-db2c0a7afcd9)

[PRR](https://github.com/danobi/prr) is a tool for doing GH code reviews in a text file (and I've [hacked together a tree-sitter grammar/Helix integration for that](https://github.com/bcspragu/tree-sitter-prr)). Setting up `delta` to use Helix as a pager, and then some minor modifications to `prr` would be a really excellent TUI/CLI code review workflow for GH. Or maybe this is an extremely niche use case :shrug:

**Comment by bcspragu** [comment]
A diffmode sounds great, though the reason I call out `delta` above is because I don't think one can get that kind of syntax highlighted diff with just a tree-sitter grammar, as it requires (at least when working from a standard diff file) assembling state across disconnected chunks of file.

Are you imagining the diff mode will have it's own system for highlighting diff chunks such that existing tree-sitter grammars work, [ala the algorithm described here in delta](https://github.com/dandavison/delta/blob/master/ARCHITECTURE.md#handling-diff-hunk-lines)?

**Comment by pascalkuthe** [comment]
> A diffmode sounds great, though the reason I call out `delta` above is because I don't think one can get that kind of syntax highlighted diff with just a tree-sitter grammar, as it requires (at least when working from a standard diff file) assembling state across disconnected chunks of file.
> 
> Are you imagining the diff mode will have it's own system for highlighting diff chunks such that existing tree-sitter grammars work, [ala the algorithm described here in delta](https://github.com/dandavison/delta/blob/master/ARCHITECTURE.md#handling-diff-hunk-lines)?

We have pur own virtual text Infrastrukturen. It's already perfectly possible to rendern multiple documents with different syntax hoghloghting interleaved.

The text rendering in helix is an order of magnitude more complex compared to a simple tool like delta.

**Comment by gyreas** [comment]
How orthogonal is it to use helix to view "rolling" log files, just like `tail -f`? Helix log will benefit directly from this since it's integrated into the editor.

**Comment by realsnick** [comment]
personally , i care less about rolling a log... im more into searching, selecting and yanking from it ... which is why id like to see such support

**Comment by aemogie** [comment]
Just wanted to mention another place where rendering raw ANSI might be useful.

When running `:sh` (The screenshot runs `:sh help help` with nushell configured as the shell)
![image](https://github.com/helix-editor/helix/assets/54412618/37bb9653-8932-4ee8-834a-395d690aaa3b)

**Comment by pascalkuthe** [comment]
closing in favor of [#2313 ](https://github.com/helix-editor/helix/issues/1976)

**Comment by johannesrave** [comment]
@pascalkuthe i just read through #2313 trying to understand how it relates to this but couldn't?

**Comment by kirawi** [comment]
https://matrix.to/#/!zMuVRxoqjyxyjSEBXc:matrix.org/$8yvUoxXuWBnYHoj-172u2Wq5x8CjVJLx_MwFjoFfkDI?via=matrix.org&via=mozilla.org&via=tchncs.de

> As soon as you are parsing/showing ansi escape does you are essentialially building a terminal emualtor anyway. So they really are the same thing. You wouldn't need to call a different pager, the builtin terminal could be used as a sort of pager directly

**Comment by k4lizen** [comment]
Same as the author, I would like to use hx as a pager for kitty, and thus don't need an integrated terminal (since kitty already has window splitting etc). While parsing ANSI codes needs to be done for an integrated terminal, surely its easier than implementing everything in regards to that?

I haven't read the whole thread there, but to me it makes sense that one would want ANSI code processing built in to helix, while an integrated terminal can be a plugin.

( i tried to open the matrix channel but my client (element) doesn't let me (: )

**Comment by kirawi** [comment]
I think one could write a CLI to strip ANSI and pipe that output to Helix. `command | strip_ansi | hx`. It could build off the https://crates.io/crates/strip-ansi-escapes crate.

Using the experimental cargo script RFC syntax:
```rs
#! usr/bin/env cargo

//! ```cargo
//! [dependencies]
//! strip-ansi-escapes = "0.2.0"
//! ```

use std::io::{self, Read, Write};

fn main() -> io::Result<()> {
    let mut stdin = io::stdin();
    let mut stdout = io::stdout();
    for line in stdin.lines() {
        stdout.write_all(line.as_bytes())?;
    }
    Ok(())
}
```

**Comment by cgahr** [comment]
There also seems to be a tree-sitter-grammar for manpages https://github.com/ribru17/tree-sitter-man that maybe could be used.

**Comment by cgahr** [comment]
> I think one could write a CLI to strip ANSI and pipe that output to Helix. `command | strip_ansi | hx`. It could build off the https://crates.io/crates/strip-ansi-escapes crate.

I think this is also not necessary. Exporting the following environment variables
```
MANPAGER = "sh -c 'col -bx | hx'";
MANWIDTH = 87;
MANROFFOPT = "-c";
```
automatically opens helix for man pages without any artifacts.

**Comment by TurtleMeds** [comment]
> closing in favor of [#2313 ](https://github.com/helix-editor/helix/issues/1976)

I think this issue needs to be reopened, as the one that it was closed in favor for is a fully different feature, although related through ANSI support. This issue is different because it is requesting support as a pager. Implementing ANSI through an integrated terminal is great, but supporting ANSI in the editor as well as the theoretical integrated terminal is what this issue is for.

**Comment by NewtonChutney** [comment]
As hx allows taking text from stdin, it should take into account programs that don't output plaintext, but only formatted text..
There should at least be some support/action to clear out known escape codes from the buffer.

**Comment by cgahr** [comment]
> As hx allows taking text from stdin, it should take into account programs that don't output plaintext, but only formatted text.. There should at least be some support/action to clear out known escape codes from the buffer.

Alternatively, you can pipe into `col -bx` to remove the unrecognized tags. For example, I'm currently using helix as my manpager via `MANPAGER = "sh -c 'col -bx | hx'";'`.

**Comment by NewtonChutney** [comment]
> Alternatively, you can pipe into `col -bx` to remove the unrecognized tags. For example, I'm currently using helix as my manpager via `MANPAGER = "sh -c 'col -bx | hx'";'`.

Thanks for pointing out alternatives. 
But the `col - bx` command doesn't completely remove coloring codes.. 

I tried with `delta --help | col -bx | hx`
There are a lot of `0m0m0m` lines.. 🥲

**Comment by cgahr** [comment]
Ah, I forgot, you have to set
```
MANROFFOPT = "-c";
```

**Comment by RoloEdits** [comment]
There is a wezterm utility that handles this as well: https://github.com/wezterm/wezterm/tree/main/strip-ansi-escapes

**Comment by k4lizen** [comment]
> Alternatively, you can pipe into col -bx to remove the unrecognized tags. For example, I'm currently using helix as my manpager via MANPAGER = "sh -c 'col -bx | hx'";' and MANROFFOPT = "-c"

Unfortunately,  then the color and bolding/underline/italics is lost, which makes it significantly less readable than when using `less` as a pager (especially with `--use-color`). Also lines are not wrapped to terminal width, which is what happens with less. I guess the mentioned man pages tree sitter could alleviate some of these concerns but certainly not all as information is simply lost when stripping.

Thing is, there doesn't need to be support for *all* ANSI escapes for this feature to be implemented, just colors, hyperlinks and a few extra things like underline, bold, italics, strikethrough. Maybe something could be inferred from the similar feature that less provides with the `--RAW-CONTROL-CHARS` option.

Related reading on man page readers specifically:
https://lobste.rs/s/fwohal/man_pages_are_great_man_readers_are

**Comment by balthild** [comment]
Hello, maintainers. Could you please reopen this issue, as it seemed to be closed in favor of an almost unrelated iusse (#2313)? Or if you had decided to reject this feature request, could you please provide an explanation on this decision?

**Comment by amandalewis1231-lgtm** [comment]
Bump
